include/forcing/NetCDFPerFeatureDataProvider.hpp

```diff
@@ -0,0 +1,464 @@
+#ifndef NGEN_NETCDF_PER_FEATURE_DATAPROVIDER_HPP
+#define NGEN_NETCDF_PER_FEATURE_DATAPROVIDER_HPP
+
+#include "DataProvider.hpp"
+
+#include <string>
+#include <algorithm>
+#include <sstream>
+#include <exception>
+
+#include <UnitsHelper.hpp>
+
+#include <netcdf>
+
+using namespace netCDF;
+using namespace netCDF::exceptions;
+
+namespace data_access
+{
+    class Catchment_Id
+    {
+        public:
+
+        Catchment_Id(std::string id) : id_str(id) {}
+        Catchment_Id(const char* id) : id_str(id) {}
+
+        operator std::string() const { return id_str; }
+
+        private:
+
+        std::string id_str;
+    };
+
+    class NetCDFPerFeatureDataProvider : public DataProvider<double, Catchment_Id>
+    {
+        
+        public:
+
+        enum TimeUnit
+        {
+            TIME_HOURS,
+            TIME_MINUETS,
+            TIME_SECONDS,
+            TIME_MILLISECONDS,
+            TIME_MICROSECONDS,
+            TIME_NANOSECONDS
+        };
+
+        NetCDFPerFeatureDataProvider(std::string input_path)
+        {
+            //open the file
+            nc_file = std::make_shared<NcFile>(input_path, NcFile::read);
+
+            //get the listing of all variables
+            auto var_set = nc_file->getVars();
+
+            // copy the variables names into the vector for easy use
+            std::for_each(var_set.begin(), var_set.end(), [&](const auto& element)
+            {
+                    variable_names.push_back(element.first);
+            });
+
+            // read the variable ids
+            auto ids = nc_file->getVar("ids"); 
+            auto id_dim_count = ids.getDimCount();
+
+            // some sanity checks
+            if ( id_dim_count > 1)
+            {
+                throw std::runtime_error("Provided NetCDF file has an \"ids\" variable with more than 1 dimension");       
+            }
+
+            auto id_dim = ids.getDim(0);
+
+            if (id_dim.isNull() )
+            {
+                throw std::runtime_error("Provided NetCDF file has a NuLL dimension for variable  \"ids\"");
+            }
+
+            auto num_ids = id_dim.getSize();
+
+            // allocate an array of character pointers 
+            std::vector< char* > string_buffers;
+
+            // resize to match dimension size
+            string_buffers.resize(num_ids);
+
+            // read the id strings
+            ids.getVar(&string_buffers[0]);
+
+            // initalize the map of catchment-name to offset location and free the strings allocated by the C library
+            size_t loc = 0;
+            for_each( string_buffers.begin(), string_buffers.end(), [&](char* str)
+            {
+                loc_ids.push_back(str);
+                id_pos[str] = loc++;
+                
+                //free(str);
```

Comment:
remove this comment if it isn't to be used

Reply:
Done, it used to be used but I found there was a `freeString()` method that should be used instead.